### PR TITLE
Allow dispatch machinery to fall back to networkx

### PR DESCRIPTION
**Pull request by eriknw**
Allow falling back to NetworkX by converting backend graphs to nx graphs. This is enabled by adding `"networkx"` to `nx.config.backend_priority`.

This PR builds off of #7496, #7499, #7502 (WIP), and #7568. I have done a little bit of refactoring, but I would like to do more refactoring to simplify the code after more PRs are merged. Feel free to review and merge the four other PRs separately, or to review this PR as a whole.

I am working on a notebook that exercises many of the branches and log/error messages, and I bet I/we will continue to refine the messages. Perhaps we could turn this into an example that gets run in CI. Hopefully the log/error messages can help make the dispatch code easier to follow (if one is comfortable reading all that text!).

CC @rlratzel

### CODE REVIEW

**Comment by eriknw** [comment]
[Here is a notebook](https://gist.github.com/eriknw/e1e9c9ceea3aa09855c8d1bc69d5d8bf
) that shows many examples using networkx dispatching. You should read this notebook when reviewing this PR. The examples are used to cover code branches and log messages top-down. Please let me know if there are more examples you wish to see, and please offer suggestions for improving messages:

https://gist.github.com/eriknw/e1e9c9ceea3aa09855c8d1bc69d5d8bf

As I said in the first post, I think we can make this much less complicated if we refactor the code to (mostly) treat networkx as any other backend, which will also set us up for backend-to-backend conversions. This notebook is also intended to be used to help with this refactor.

**Comment by eriknw** [comment]
The refactor to generically allow fallback to backends ("networkx" by default) and, soon, backend-to-backend conversions is complete. This notebook shows the updated log and error messages: [fallback_to_nx2.ipynb](https://gist.github.com/eriknw/e1e9c9ceea3aa09855c8d1bc69d5d8bf#file-fallback_to_nx2-ipynb) and can be compared to the first notebook in the gist, which is from an earlier version of this PR.

There are some remaining TODOs needed to get this over the finish line, but the big push is finished, and curious reviewers can begin looking at this PR. I think the code in `__call__` is cleaner and clearer.

I may want to discuss some things about this PR in the weekly dispatch meeting. In particular, @jim22k helped come up with the design and behavior of the new `nx.config.backend_fallback` config.

**Comment by eriknw** [comment]
This is ready for review :rocket:

**Comment by mjschwenne** [comment]
The examples were very helpful to understanding the proposed behavior. That is the role of the `backend_fallback` config option? Is this any different than putting `"networkx"` as the last position in the `backend_priority` list?

**Comment by eriknw** [comment]
> What is the role of the `backend_fallback` config option? Is this any different than putting `"networkx"` as the last position in the `backend_priority` list?

Great question @mjschwenne! As we've developed dispatching, we've come up with a few guidelines for when and how to do dispatching and conversions. For example, the more explicitly one uses a backend, then the more hard we should try to use that backend. Thus, we have:
1. If `backend=backend_name` is given then use that backend (and fail if can't)
2. If you pass in a backend graph to an algorithm, then _try_ to use that backend
3. If you pass in a NetworkX graph to an algorithm, then try to convert it to use backends in `backend_priority`

So if we treat `"networkx"` as "just another backend"--and want to someday allow other backends to behave like `"networkx"` so it's not a "special" backend--then rules (2) and (3) are in direct conflict. (2) indicates that if you use a graph from backend `"networkx"`, then you should use that backend, but this is not the behavior we developed nor want.

In other words, backends in `backend_priority` are "sticky" and won't be converted (even to higher priority backends) if they implement a function. Putting `"networkx"` in `backend_priority` would make it "sticky", which we don't want.

`backend_fallback`, as the name implies, is intended for _falling back_. We will try to convert and use these backends _after_ trying to use higher priority backends (backend from inputs, and backends in `backend_priority`). Backends in `backend_fallback` _are not_ "sticky", so they will be attempted to be converted to use backends listed in `backend_priority` if possible.

I know this is a little nuanced and complex, but I think it should typically "just work" and enable us to support patterns we'll want to use in the future. It also gives control to users to configure as best suites their needs.

We don't strictly need `backend_fallback` as a user-facing configuration for 3.4, because we don't yet have backend-to-backend conversions, which is when this config would be generally useful. We could hide it, or come up with a simpler configuration for convenience such as `fallback_to_nx` or whatever.

**Comment by dschult** [comment]
So, if I understand right, we now have two related configuration settings: `backend_priority` and `backend_fallback`. The priority says what order to try backends when a NXGraph is the input. The fallback is what backend to try if a backend conversion fails. But this probably isn't a correct reading of your description. 

Does the fallback option occur when **any** of the priority backends can't convert? Or after **all** of the priority backends can't convert?

And I'm not sure what you mean by 'sticky': "backends in backend_priority won't be converted if they implement a function."  doesn't make sense to me... is there a word missing?

**Comment by eriknw** [comment]
> So, if I understand right, we now have two related configuration settings: `backend_priority` and `backend_fallback`.

Correct.

> The priority says what order to try backends when a NXGraph is the input.

Typically, yes, but done in a way that does not treat `"networkx"` as a special backend.

> The fallback is what backend to try if a backend conversion fails. But this probably isn't a correct reading of your description.

Depends what you mean by "backend conversion fails". "Fail" here covers the inability to convert to a higher priority backend (i.e., `can_convert` is false), conversion fails (i.e., `NotImplementedError`), and a higher priority backend (including the backend of the input) doesn't implement a function or `should_run` is false or `can_run` is false or the backend function raises `NotImplementedError`.

So, basically, yeah, fallback happens last after everything else fails.

> Does the fallback option occur when any of the priority backends can't convert? Or after all of the priority backends can't convert?

I'm not sure I understand. All backends in `backend_priority` are tried before those in `backend_fallback`.

> And I'm not sure what you mean by 'sticky': "backends in backend_priority won't be converted if they implement a function." doesn't make sense to me... is there a word missing?

Let's show with an example with backends `"spam"` and `"eggs"` with backend graphs `G_spam` and `G_eggs`.
```python
nx.config.backend_priority = ["spam", "eggs"]
nx.pagerank(G_spam)  # Uses "spam" backend
nx.pagerank(G_eggs)  # Uses "eggs" backend
```
I called these "sticky" (a sloppy term) to indicate that they will try first to run _without_ converting.

In contrast, if only `"spam"` was in `backend_priority`, then `G_eggs` would try to run with `"spam"`:
```python
nx.config.backend_priority = ["spam"]
nx.pagerank(G_spam)  # Uses "spam" backend
nx.pagerank(G_eggs)  # Tries to convert and use "spam" backend, then would try "eggs" backend
```

I believe typical usage, especially typical _introductory_ usage, will only need to set `backend_priority` to a single backend and leave `backend_fallback` untouched. We currently do not support backend-to-backend conversions generically--we only support converting to and from networkx. My hope is that this will "just work" and feel intuitive when used for actual use cases.

**Comment by dschult** [comment]
Thanks for that description -- still digesting it.

What is the difference between the backend in "fallback" and having that backend in the last spot of the "priority"?

**Comment by dschult** [comment]
> Thus, we have:

> 1. If backend=backend_name is given then use that backend (and fail if can't)
> 2. If you pass in a backend graph to an algorithm, then try to use that backend
> 3. If you pass in a NetworkX graph to an algorithm, then try to convert it to use backends in backend_priority

I have trouble with Rule 2. It seems to assume that every backend has a single backend graph type. And every graph type has a single backend associated with it.  We ran into this already with nx-parallel and again with nxadb where the graph class structure is a networkx graph. We got around that. But perhaps we should consider removing that assumption. Sometimes we are trying to make backends replace functions, and other times we are trying to make them replace graph classes, and often we are doing both. Is there an interface that would separate those two approaches?

Maybe each backend should support a tuple of graph classes that it know about and supports. And each graph class can hold a tuple of graph classes it can convert to. Then we remove Rule 2 and adapt Rule 3 to march along the backend_priority looking for a backend that supports a graph class that the input graph can convert to.  (Note that "supports a graph class" basically means that the backend can convert the incoming graph class to something it can use.)

**Comment by dschult** [comment]
Sorry for multiple posts instead of one...

> nx.config.backend_priority = ["spam"]
> nx.pagerank(G_spam)  # Uses "spam" backend
> nx.pagerank(G_eggs)  # Tries to convert and use "spam" backend, then would try "eggs" backend

Does `tries to convert and use "spam" backend` mean check if the spam backend function works with G_eggs as input? (that is, does this happen by an indicator on the backend, or by trying to run the function and looking for exceptions?)

**Comment by eriknw** [comment]
Thanks for engaging and asking great questions @dschult! Dispatching with conversions is a delightfully delicate and complex topic, so I think it's worth diving into.

Before I go into replies, I want step back and give perspective for this PR. This PR contains the changes in #7496, #7499, #7502, and #7568, each of which can be reviewed and considered independently. In addition to those changes, this PR adds the ability to fall back to the default networkx implementation (it also fixes a couple bugs thanks to the refactor and test notebooks). On its own, I think this is a nice and non-controversial new feature. As I mentioned in https://github.com/networkx/networkx/pull/7585#issuecomment-2302465175, we can delay exposing `backend_fallback` config until after 3.4 to give us more time to deliberate and explore alternatives. If there's appetite, I'd like to continue discussing `backend_fallback` _right now_ as well, but if we can't come to consensus in time for 3.4 we should just hide it for now and continue later.

@jim22k helped come up with the new design, configs, and behaviors in this PR, so maybe it would be helpful if he could try to explain things too. I recall @jim22k also found the term "sticky" confusing when I used it trying to explain behavior to him. He's been closely involved in designing dispatching in networkx since the very beginning.

Also, several of us will be in person next week, and I hope we can do a "deep dive" together and maybe come up with a document or some such explaining things. I am particularly interested to get @seberg's take!

I found one of your ideas very interesting: allowing `__networkx_backend__` to be a tuple of backends.

I'll reply more later to try to respond to your questions.

**Comment by eriknw** [comment]
I removed `nx.config.backend_fallback` for now and simply default to falling back to networkx. I'd like to try to get this PR reviewed and in soon because it's a much-needed refactor (with some bugfixes), and then we can discuss and make new PRs for how to configure and maybe generalize fallback behavior. Does this sound good?

@dschult, thanks again for taking a look and asking good questions. I will still try to answer, but it may take me some time.

**Comment by jim22k** [comment]
> I have trouble with Rule 2. It seems to assume that every backend has a single backend graph type. And every graph type has a single backend associated with it. 

I think @dschult brings up a good point. I was definitely thinking of a 1:1 relationship between graph type and graph library.

I would prefer the behavior to be:

1. If `backend=backend_name` is given then use that backend (and fail if can't)
2. Step through the `backend_priority` entries -- if backend can handle the graph without conversion, call it
3. Step through the `backend_priority` entries -- if backend can handle the graph **_with_** conversion, call it
4. Step through all registered backends (in alphabetical order) -- if any can handle the graph without conversion, call it
5. Step through the `backend_fallback` entries -- if backend can handle the graph **_with_** conversion, call it

This removes the need to lookup a backend based on the graph type. It is deterministic. It provides the user with several ways to specify their preferred backend while minimizing unnecessary graph conversions.

The `backend_priority` specifies both the backend preference as well as authorizes networkx to do graph conversions. The `backend_fallback` is only used as a last resort when conversion is required.

@eriknw Feel free to push back against any of this. You have spent a lot more time thinking about this than I have. But I'm hoping this suggestion feels simple, understandable, and useful.

**Comment by rlratzel** [comment]
I'm glad we're having this discussion.  I agree with many of the points brought up by @dschult and @jim22k, but I'd take it a bit further:
1. If `backend=backend_name` is given then use that backend (and fail if can't)
2. Step through the `backend_priority` entries -- if a backend cannot handle the input graph, it raises `NotImplementedError` and the next backend in the list is tried. Maybe "networkx" is an implied/default last entry in the list.

And that's it.  Since "many" backends (3 of the 4 publicly-known backends that I'm aware of) have graph types that are either NetworkX graph types or drop-in/duck-type compatible NetworkX graph types, I'm thinking it's feasible to just formalize that as a requirement for backends.  Supporting a library's types seems like a reasonable requirement for a backend to said library IMO.  Furthermore, if a backend wishes to be used for a type other than its own (including NX types), it's free to convert it itself (using either `from/to_networkx`, or any other means that makes sense for them), and cache the conversion as a `__networkx_cache__` entry if it wishes and if that dunder is supported by that type.  This gives backends the opportunity to optimize as they see fit ("backend-to-backend" conversion). 

This seems like it would simplify core NetworkX dispatching tremendously - NetworkX doesn't care about backend types, conversions, and caching anymore, it just passes graphs to functions in priority order.  Of course, it would require changes to some backends and NetworkX, but I think it would simplify the code, docs, tests, etc. in the long run.  We'd have to get there from here incrementally in a backwards-compatible way.

But I don't mean to block this PR if the community wants it in, plus this PR does solve problems faced today (for `nx-cugraph` use cases, possibly others), so these suggestions can be applied to a later discussion for future PRs if necessary.

**Review by eriknw — COMMENTED** [review]
Thanks everybody for the discussions and sharing your ideas! I'm glad this provoked some thought. Hearing your suggestions gives me hope we may be able to find a suitable path forward to continue improving dispatching beyond networkx v3.4.

I would like to focus on being practical right now and try to get this PR reviewed, accepted, and merged. I've spent _a lot_ of time using this branch the last few weeks and have continued to refine log messages and fix bugs. I think this is a very important PR to get in.

Please let me know if there is more I can do to help make this more reviewable. As it is a combination of 4 other PRs _and_ a refactor, it has gotten larger than is typically recommended for an PR on an open source package with volunteer reviewers. I will do what I can to help.

I am working on adding better ways to test the dispatch machinery, and I plan to port the example notebooks from the first comment as tests. These could be part of this PR or as a new PR. Either way, I plan to use them to demonstrate continuity and changes pre- and post- this PR.

I left a few comments to draw attention to a few important things.

File: networkx/utils/backends.py
Comment (dschult) [inline networkx/utils/backends.py:2259]: Should `val` be called `graph` here?
If `val` is in fact a better name (and it could be...) then we should probably change the name `val` in line 2290-2291.

**Review by dschult — COMMENTED** [review]
I think I have gone through the aspects of this. I was able to use the draft PR #7499 to get the ideas for that in a smaller setting. So thanks for that!

I like the refactoring in general and the updates on the config systems. New names for the difference `val` variables are minor but nicely helpful too.

I'm not sure I like the idea of automatically converting other backends to networkx graphs (which I think is the meaning of the phrase "fall back"... it's not just that we use networkx functions, its that we convert_and_run other graph objects to nx graphs and then run nx functions on them). But I'm still learning. What is the expected use-case here and how can we avoid surprises? It's a fair amount of extra cruft compared to just raising an exception and having the user replace `G` with `nx.Graph(G)`.

I support backend-to-backend conversions, but I don't like the approach of automatically choosing which to which. It's too implicit. What are the advantages of automatic conversion over user-directed explicit conversion?

I **do** think we should allow the keyword arg `backend="networkx"`. That is, this is the explicit way to say we want the original networkx functionality even if we have `backend_priority` set.

Can we update the Backends docs page to reflect these changes? For example, it would be nice to tell readers interested in details that `backend_info` holds as keys the backends that can be used. There are places that say ``backend_info` is optional. Is that still true? (I think so) I guess I don't understand whether and why `backend_info` holds as keys the backend names instead of whereever it used to be... (which was `backends`?)  We should also reproduce the Rules for choice of backend from comments in the code to the docs. 

I'm beginning to think that type-based dispatching is pretty restrictive. In some sense, I'd like to see the code that is dispatched to be one choice and the object used by that code as a separate choice. So libraries could support more than one type (by converting and caching internally or by making code that works across types), and **also** more than one library can support a specific type (like a number of libraries which work directly with the nx graphs).  But I haven't thought about how to implement that.

To finish up here with this comment:
- what aspects of this do you see as needed for v3.4? 
- Do we need the conversion of other types to networkx graphs upon fallback? What does this feature enable? 
- Is this going to lock us into the five groups of priorities described in lines 1100-1107?  It is very appealing to just have `backend_priority`, and go down that list and pick the first one that doesn't signal `NotImplemented`.  Backend-to-backend conversion can still happen when the dispatched-to library can handle/convert the incoming graph type. And if it can't then return `NotImplemented` (or raise, or whatever signal works best).

File: networkx/algorithms/components/tests/test_connected.py
Comment (eriknw) [reply networkx/algorithms/components/tests/test_connected.py:96]: `nx_loopback` is only registered as a backend when testing backends.

File: networkx/utils/tests/test_backends.py
Comment (eriknw) [reply networkx/utils/tests/test_backends.py:98]: Harder for us NX maintainers. Improving testing of dispatch machinery is high priority and I'd like us/me to work to improve it in the near future. In the mean time, I think we should be happy that we have this test even if it's a bit awkward.

File: networkx/utils/backends.py
Comment (eriknw) [reply networkx/utils/backends.py]: I added a comment. It's strictly for aesthetics so it's easier to search for.

File: networkx/utils/backends.py
Comment (eriknw) [reply networkx/utils/backends.py]: I have no preference. Changed to `graph`

**Comment by eriknw** [comment]
> I think I have gone through the aspects of this. I was able to use the draft PR #7499 to get the ideas for that in a smaller setting. So thanks for that!

You bet, thanks for taking a close look!

> I like the refactoring in general and the updates on the config systems. New names for the difference `val` variables are minor but nicely helpful too.

:+1:. The PR has been updated with your suggestions.

> I'm not sure I like the idea of automatically converting other backends to networkx graphs (which I think is the meaning of the phrase "fall back"... it's not just that we use networkx functions, its that we convert_and_run other graph objects to nx graphs and then run nx functions on them). But I'm still learning. What is the expected use-case here and how can we avoid surprises? It's a fair amount of extra cruft compared to just raising an exception and having the user replace `G` with `nx.Graph(G)`.

As we recently discussed, I change the default of `nx.config.fallback_to_nx` to False. I also updated the config docstring to try to make it more clear. Falling back to networkx implementation allows `nx-cugraph` to achieve "zero code change" for virtually _all_ of networkx, although this comes at the cost of creating the networkx graph data structures (`_adj`, etc.). Conversion is optional; we call the backend interface `convert_to_nx` with the function that it's being converted for, and the backend may choose to not convert if the graph is already duck-type compatible as a networkx graph (this is what `nx-cugraph` does).

> I support backend-to-backend conversions, but I don't like the approach of automatically choosing which to which. It's too implicit. What are the advantages of automatic conversion over user-directed explicit conversion?

Gotcha. I think this is a conversation for later. This PR does not enable generic backend-to-backend conversions, and I agree it would be great to allow user-directed explicit conversions.

> I **do** think we should allow the keyword arg `backend="networkx"`. That is, this is the explicit way to say we want the original networkx functionality even if we have `backend_priority` set.

Great! I like it too.

> Can we update the Backends docs page to reflect these changes? For example, it would be nice to tell readers interested in details that `backend_info` holds as keys the backends that can be used. There are places that say `backend_info` is optional. Is that still true? (I think so) I guess I don't understand whether and why `backend_info` holds as keys the backend names instead of whereever it used to be... (which was `backends`?) We should also reproduce the Rules for choice of backend from comments in the code to the docs.

I've done my best to update the docs to reflect these changes. I don't think most users should be using `backend_info`; `nx.config.backends` has the installed backends.

> I'm beginning to think that type-based dispatching is pretty restrictive. In some sense, I'd like to see the code that is dispatched to be one choice and the object used by that code as a separate choice. So libraries could support more than one type (by converting and caching internally or by making code that works across types), and **also** more than one library can support a specific type (like a number of libraries which work directly with the nx graphs). But I haven't thought about how to implement that.

I'm similarly surprised we've been able to get as far as we have with our current dispatching! I like this chain of thought, so let's try to continue it later.

> To finish up here with this comment:
> 
> * what aspects of this do you see as needed for v3.4?

All of it.

It has all the things we need, and some of the things we want. There are bug fixes, and log messages are greatly improved. It is systematically tested. The code should be easier to maintain and more adaptable to change. It does not lock us into an approach for the "backend-to-backend" story.

> * Do we need the conversion of other types to networkx graphs upon fallback? What does this feature enable?

It allows for a smoother experience when users have backend graph objects. Just like they can opt in to automatically converting a networkx graph to a backend graph, this allows the reverse. Also, as I said above, it supports the "no code change" use case for nx-cugraph.

> * Is this going to lock us into the five groups of priorities described in lines 1100-1107?  It is very appealing to just have `backend_priority`, and go down that list and pick the first one that doesn't signal `NotImplemented`.  Backend-to-backend conversion can still happen when the dispatched-to library can handle/convert the incoming graph type. And if it can't then return `NotImplemented` (or raise, or whatever signal works best).

Not at all! In fact, I think it's the opposite: now that we've identified the five groups, it's much easier to change. The refactor is much more adaptable. Currently, we do not do backend-to-backend stuff. My bad for discussing it so much. Should I delete some code comments that discuss future possibilities?

Finally, two broader issues have come up:
1. we would like better testing of dispatch machinery.
2. we would like better maintainer documentation of dispatching, perhaps as a NXEP.

These are true regardless of this PR. I agree they are high priority, and I hope we can address them soon.

File: networkx/algorithms/link_analysis/tests/test_pagerank.py
Comment (eriknw) [reply networkx/algorithms/link_analysis/tests/test_pagerank.py:87]: Yeah, that's exactly right. We include this one still for sanity. It's here, because it's where @jim22k added it ~2.5 years ago :joy:. Yeah, it would be nice to move it to `test_backends.py`... hopefully part of efforts to improve dispatching testing.

File: networkx/utils/backends.py
Comment (dschult) [inline networkx/utils/backends.py:74]: To avoid confusion about the term "fall back", perhaps we could state that conversion happens.(?)
```suggestion
it will convert the input graphs to NetworkX classes and run the default
networkx implementation. And, if the function mutates the input graph
```

Comment (eriknw) [reply networkx/utils/backends.py]: That's better, thanks!

**Review by MridulS — APPROVED** [review]
I was about to add a bunch of littered comments all around but decided not to do that.

> we would like better testing of dispatch machinery.
> we would like better maintainer documentation of dispatching, perhaps as a NXEP.

Most of the comments were around these 2 points (as pointed out by @eriknw) that we need to work on in the near future with dispatch machinery.

Let's get this in!

**Review by dschult — APPROVED** [review]
I, like @MridulS, have a smattering of changes, mostly about docs and tests. I will put them into a separate PR(s). 
I'm still looking for a simple explanation of how the backends are determined, and I am pretty sure the current description is not complete. But with this PR the code is easier to read and the decision framework described in the comments seems like it could be turned into some good docs.

So: "Once more unto the breach, dear friends" 
Let's merge and pick up any stray pieces as we find them.

**Pull request merged**
